### PR TITLE
Adds IPv6 prefix to VPC objects

### PR DESCRIPTION
**Pull request by bnaecker**
- Adds an optional IPv6 prefix to the request to create a VPC. Validated
  if provided, otherwise a random RFD 4139 address is created.
- Adds the prefix to the database and views
- Adds some integration tests verifying this behavior

### CODE REVIEW

File: nexus/src/defaults.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    use rand::Rng;
+    let mut bytes = [0u16; 8];
+    bytes[0] = 0xfd00;
+    rand::thread_rng().try_fill(&mut bytes[1..4]).map_err(|_| {
+        external::Error::internal_error(
+            "Unable to allocate random IPv6 address range",
+        )
+    })?;
+    Ok(Ipv6Net(Ipv6Network::new(Ipv6Addr::from(bytes), 48).unwrap()))
+}
```
Comment (david-crespo) [inline nexus/src/defaults.rs:65]: maybe this makes no sense, but I find it a little weird to think of a random value as a default value

Comment (bnaecker) [reply nexus/src/defaults.rs:65]: See [RFD 21 section 3.4.2](https://rfd.shared.oxide.computer/rfd/0021#_vpc_apis). Clients can specify a prefix if they wish, or a VPC is assigned a default. This is a [unique local address](https://en.wikipedia.org/wiki/Unique_local_address), defined in [RFC 4193](https://www.ietf.org/archive/id/draft-hain-ipv6-ulac-02.txt). It's just part of the spec.

Comment (david-crespo) [reply nexus/src/defaults.rs:65]: Sorry for being unclear — I was talking about the creation of this file called `defaults.rs` as an organizational nitpick, not to the use of this function. I see now that it exists in order to be able to share common code between `model.rs` and `nexus.rs`. I'm not sure `defaults` will hold up as a category of things that belong together but it's not very important.

Comment (bnaecker) [reply nexus/src/defaults.rs:65]: Ah, I see. Yeah, some of this is shared in a few places. I intended this to be a location for common default values (or ways to generate them) that are part of the public API, but I'm not dogmatic about where it goes.

File: nexus/src/external_api/params.rs [hunk 2]
```diff
@@ -107,6 +107,7 @@ pub struct InstanceMigrate {
 pub struct VpcCreate {
     #[serde(flatten)]
     pub identity: IdentityMetadataCreateParams,
+    pub ipv6_prefix: Option<Ipv6Net>,
```
Comment (zephraph) [inline nexus/src/external_api/params.rs:110]: Is `Ipv6Net` the correct type to use here? I'd guess any `IPv6` prefix is a proper `IPv6Net` interface but the inverse isn't true. If we expressly created an `Ipv6Prefix` type we could give more expressive results in the API to inform its usage. That'd also move the conversion logic out of the VPC model and isolate it to this type.

This isn't blocking.

Comment (bnaecker) [reply nexus/src/external_api/params.rs:110]: Thanks for taking a look @zephraph! If I understand your question, you're wondering if we might allow specifying _only_ the 48-bit prefix that unique local addresses share? Or even the 40 random bits, as the first 8 are required to be `fd`.

If so, that'd probably be more confusing than helpful, I'd expect. Networking folks would likely expect to work with the full, standard subnet mask notation, and the type wouldn't really be used anywhere but this specific API call.

**Pull request merged**